### ISSUE TITLE
Add UR2.0 QR codes

### ISSUE THREAD

**Issue by Jakob-1870** [spesmilo/electrum#7905]
UR 2.0 is becoming the standard for QR PSBT's. It's implemented in almost all software and hardware wallets that support QR PSBT's.
https://github.com/BlockchainCommons/Research

Electrum should add support for it.

**Comment by SomberNight** [comment]
> https://github.com/BlockchainCommons/Research

There's a lot of stuff there... Which "spec" in particular are you referring to? Is it this one?
https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md

> UR 2.0 is becoming the standard for QR PSBT's. It's implemented in almost all software and hardware wallets that support QR PSBT's.

Which other software/hardware/services support it?

**Comment by Jakob-1870** [comment]
Yes that's it.

Bluewallet, Sparrow, Specter, Keystone, Passport, Gordian, Krux, Seedsigner all use UR2.0.

**Comment by stepansnigirev** [comment]
Two python implementation that might be interesting for you:
From Foundation Devices: https://github.com/Foundation-Devices/foundation-ur-py
Implementation in Specter-DIY: https://github.com/diybitcoinhardware/f469-disco/tree/master/libs/common/microur

**Comment by massmux** [comment]
Sorry for duplicating the request i did not get aware of this one. But yes this would be important i think for all the community

**Comment by SomberNight** [comment]
Some previous discussion re the custom format we have been using since 2015: https://github.com/spesmilo/electrum/issues/5746
In short, when implementing PSBT support in 2019 we decided to keep using the base43 encoding inside alphanumeric-mode QR codes, instead of using either base64 text inside binary-mode QR codes or raw bytes inside binary-mode QR codes, due to maximising payload size and easy-copy-pasteability.

---

Anyway, if this BCR-2020-005 format has traction, we could have it too. PR welcome. It probably needs some GUI-specific work, but getting it to work with the Qt GUI would be a good starting point.

**Comment by ecdsa** [comment]
I think it would be a pity if we were forced to adopt an animated qr code standard just because some spec is feature bloated.
Has someone taken the time to evaluate that?

**Comment by massmux** [comment]
actually sparrow and specter desktop use such a format. maybe it would be important electrum's users also can share same standard. imho

**Comment by Medgeina** [comment]
I too think that the use of QRcodes would be very practical. There would be less room for error, it takes away the fragility caused by the use of USB or SDcards... also, if everyone used the same standard, it would be easier to use different wallets/devices for broadcasting

**Comment by massmux** [comment]
@SomberNight can you please write here in this thread, where a dev willing contribute, can find appropriate code parts to update. This may be makes things easier for finding the help to include this change

**Comment by amarett0** [comment]
Is there any progress with this?
I think the Jade hardware wallet uses this system

**Comment by tadeubas** [comment]
Krux also uses this system of animated QRCodes (UR types) and it doesn't work with Electrum, but it works with BW, Sparrow and others... the implementation is here (Python): https://github.com/selfcustody/urtypes

**Comment by SomberNight** [comment]
> the implementation is here (Python): https://github.com/selfcustody/urtypes

Is that really implementing the same thing? Looks unrelated to me.
The code linked in https://github.com/spesmilo/electrum/issues/7905#issuecomment-1338994775 look much more promising.

**Comment by tadeubas** [comment]
> > the implementation is here (Python): https://github.com/selfcustody/urtypes
> 
> Is that really implementing the same thing? Looks unrelated to me. The code linked in [#7905 (comment)](https://github.com/spesmilo/electrum/issues/7905#issuecomment-1338994775) look much more promising.

Yes you are right, the other code is just a helper to determine the type of the UR... Krux also uses that code as a simple fork from Foundation Devices: https://github.com/selfcustody/foundation-ur-py

**Comment by tadeubas** [comment]
Maybe this project in Python can help Electrum achieve Multiframes QRCodes too:
https://github.com/pythcoiner/SeedQReader

**Comment by Ciugam** [comment]
Are there any news on this topic?
I think it's important to implement, thanks to seedsigner and jade's QRs features

**Comment by dangarthwaite** [comment]
I'm willing to pay a bounty for this feature.  It would allow old phones, with internet disabled, to become multisig participants.

**Comment by liebregris** [comment]
> I'm willing to pay a bounty for this feature. It would allow old phones, with internet disabled, to become multisig participants.

A bit off topic, just out of curiosity: do you mean use old phones as signers using electrum, may I assume electrum android app?

**Comment by dangarthwaite** [comment]
@liebregris Yes, that's something I can see value in.  Everyone has a junk drawer full of old phones, most of them Android.

**Comment by Ericrowlowland** [comment]
Tweeting won't solve the problem it will only bring fake recommendations
and probability of getting duped again. I have experienced the same problem
but I got help from a legit computer expertise on Instagram
https://instagram.com/kay_cyberhack?igshid=NTc4MTIwNjQ2YQ==
Reach out to him for help.

**Comment by massimopavoni** [comment]
Are there any updates on this? Or is there a sort of workaround, like saving a psbt and then using another tool to generate the codes?

**Comment by accumulator** [comment]
@massimopavoni we will update this issue if there is any progress here

**Comment by tiagodread** [comment]
+1 For supporting this feature, would help a lot.

**Comment by LeoDroves** [comment]
This feature would be very good to setups that uses TailOS+Electrum offline with PSBT on BlueWallet Watch Only

**Comment by qtipbera** [comment]
Seed Signer cannot work with Electrum without UR2.0 QR codes for PSBTs. This is unfortunate because Electrum seems superior to Sparrow.

**Comment by hectorchu** [comment]
https://youtu.be/1OJae_odl_I

EDIT(sombernight): ^ video showing implementation in Electrum-LTC.

**Comment by SomberNight** [comment]
@hectorchu is it mostly this commit? https://github.com/ltc-electrum/electrum-ltc/commit/4d4c638e19b4ec4500a419550a53b9b446ea447e
Though it has some other stuff mixed in there.
Looks like it has no external dependencies, which is nice!
Note that we would also want to support this on Android.

**Comment by hectorchu** [comment]
Yes that commit adds UR codes among Cupcake support. Should be qrcodewidget, camera_dialog, and the ur folder. Not tested on Android.

**Comment by SomberNight** [comment]
Ok so to sum up, atm we use a somewhat custom format to display txs as QR codes.
Since we started doing that, apparently two competing "standards" emerged (BC-UR 2.0 and BBQr).
Out of the two, BC-UR 2.0 looks better adopted, and I don't see a clear advantage to BBQr.

It is "easy" to be able to read (input) multiple formats, but it is a huge mess to be able to display (output) multiple formats.
I don't want to offer a choice between QR code formats when displaying, I don't want to ask the user.
(Maaaaaybe there could be a hidden config option for it......)

So, I think we could switch to using BC-UR 2.0 for psbts/complete-txs.
For output: we would always display BC-UR 2.0.
For input: we could support both BC-UR 2.0 and our existing custom format.
Basically we would almost just hard-switch.

-----

Given @hectorchu's impl in https://github.com/ltc-electrum/electrum-ltc/commit/4d4c638e19b4ec4500a419550a53b9b446ea447e, significant parts of this are readily available. Parts of that commit should be extracted.

However, Android support is missing, which is a prerequisite - we cannot leave Android QR codes incompatible with desktop. Unfortunately the Android part is non-trivial, as atm the QR scanner and camera is in Java land, however all the BC-UR logic would be in python. We need the BC-UR logic in python as e.g. the desktop GUI would definitely use that. If the BC-UR logic was duplicated in Java, the Android QR scanner could easily use that, but that sounds really ugly. The alternative however involves crossing the python-java boundary for this logic, which might not be that simple (but maybe it is? I haven't had time for this yet).

see https://github.com/spesmilo/electrum/blob/f8fc2b63e321078f59a45c3565e6043fbdec2d5d/electrum/gui/qml/qeqrscanner.py#L52

So yeah, someone needs to pick this up.

**Comment by accumulator** [comment]
We still have the code and UI for QR scanning and parsing in QML and python, this was disabled due to instability of using the camera in Qt<=6.4. If Qt>6.4 has this fixed, we could revert to that after upgrading Qt and take the python code for parsing BC-UR.

**Comment by hectorchu** [comment]
I don't think just doing the simple thing and using a Java UR library is 'ugly'. There's a good one (https://github.com/sparrowwallet/hummingbird) that's written by Sparrow. Can you please explain why doing that is "really ugly"? As a fellow developer and having thought about this for ten minutes I'm struggling to see why the opposition to this approach.

**Comment by SomberNight** [comment]
The ugly part is having the same logic duplicated in python and in java. Especially if it means adding more dependencies (and transitive deps).

I think it should very much be feasible to pass around the QR code scan results and a "keep scanning" vs "stop scanning" signal using e.g. [Intents](https://developer.android.com/reference/android/content/BroadcastReceiver?hl=en) between the [`SimpleScannerActivity`](https://github.com/spesmilo/electrum/blob/d112653eaa7686ea8eaa97c219645a66f0ddd400/electrum/gui/qml/java_classes/org/electrum/qr/SimpleScannerActivity.java) and the main [`PythonActivity`](https://github.com/spesmilo/electrum/blob/d112653eaa7686ea8eaa97c219645a66f0ddd400/electrum/gui/qml/qeqrscanner.py#L50-L59).

That hummingbeard library also depends on a cbor java library. That's more logic that we ~have in python. So then this would mean adding two new dependencies we have to review, keep updated, and reproducibly build.
And someone would still have to write some glue code to make it work. Perhaps it's roughly the same amount of code that we are missing for the python Intent-based communication approach - we might as well just have that and maintain that.

It's better if the same logic is used both on desktop and android.

We are not optimising for one time implementation cost, but rather for long term maintenance and ease of understanding. It's simple to keep adding stuff but it is not maintainable.

**Comment by hectorchu** [comment]
Here is the implementation of a scanner that decodes UR codes, in Android:

https://github.com/spesmilo/electrum/compare/master...ltc-electrum:electrum-ltc:ur_scanner

**Comment by SomberNight** [comment]
That looks very promising!  It's even simpler than what I described with Intents, very nice.

Could you please open a pull request from a branch that has both desktop and android changes?

**Comment by hectorchu** [comment]
https://github.com/spesmilo/electrum/pull/10518

**Comment by ecdsa** [comment]
> The ugly part is having the same logic duplicated in python and in java. Especially if it means adding more dependencies (and transitive deps).

I agree. Since we need the python logic for the desktop app, I too believe that it is better not to use the java library. 
It makes desktop and android consistent.
